include/kernel.h

```diff
@@ -1417,6 +1421,314 @@ struct k_timer {
+ * of the ready list.
+ *
+ * This is invoked from the platform-specific timer module interrupt
+ * handler when the system cycle counter reaches a value relevant to
+ * the alarm infrastructure.
+ *
+ * @note This is not user API.
+ *
+ * @retval negative if a non-empty ready list was left unchanged.
+ *
+ * @retval zero if there are no ready alarms.
+ *
+ * @retval postive the number of alarms added to the ready list.
+ */
+int k_alarm_split_(u32_t now);
+
+/** Walk the set of ready alarms invoking each one's handler.
+ *
+ * This is invoked from the platform-specific timer module interrupt
+ * handler after the system clocks have been updated and after any
+ * k_timer processing has completed.  It iteratively invokes the
+ * handlers for each ready alarm in order of scheduling within
+ * deadline.
+ *
+ * @note This is not user API.
+ */
+void k_alarm_process_ready_(void);
+
+/** Schedule the alarm to fire at the corresponding deadline.
+ *
+ * Normally an alarm can be scheduled only when its state is
+ * #K_ALARM_UNSCHEDULED, #K_ALARM_ACTIVE, or #K_ALARM_CANCELLED.  The
+ * @p flags parameter can be used to allow scheduling when in state
+ * #K_ALARM_SCHEDULED under certain circumstances.
+ *
+ * @param alarm the alarm to be scheduled.  This must be in
+ *  state.
+ *
+ * @param deadline the low 32 bits of the system clock time at which
+ * the alarm should be scheduled.  If this is not less than 2^31 ticks
+ * in the future the deadline will be interpreted to have passed, and
+ * the alarm will immediately transition to the READY state.
+ *
+ * @param flags options that affect the scheduling behavior, including
+ * `K_ALARM_FLAG_ERROR_IF_LATE`, `K_ALARM_FLAG_REPLACE`, and
+ * `K_ALARM_FLAG_IF_SOONER`.
+ *
+ * @retval positive if the alarm has been scheduled to fire in the future.
+ *
+ * @retval zero if the alarm has been added to the ready list.
+ *
+ * @retval -EINVAL if @p alarm is null.
+ *
+ * @retval -EINVAL if the alarm state is corrupted.
+ *
+ * @retval -EBUSY if the alarm is not in UNSCHEDULED, ACTIVE, or
+ * CANCELLED state.
+ */
+__syscall int k_alarm_schedule(struct k_alarm *alarm,
+			       u32_t deadline,
+			       u32_t flags);
+
+int z_impl_k_alarm_schedule(struct k_alarm *alarm,
+			    u32_t deadline,
+			    u32_t flags);
+
+/** Cancel a scheduled or ready alarm.
+ *
+ * If a non-error result is returned the registered cancel handler
+ * will have been invoked, and if it did not reschedule the alarm the
+ * alarm is left unscheduled.
+ *
+ * @retval positive if the alarm was scheduled but had not reached its
+ * deadline at the point the cancellation was requested.
+ *
+ * @retval 0 if the alarm had reached its deadline but the callback
+ * had not been invoked at the point where the cancellation was requested.
+ *
+ * @retval -EINVAL if the alarm was in any other state than SCHEDULED
+ * or READY when the cancellation was requested.
+ */
+__syscall int k_alarm_cancel(struct k_alarm *alarm);
```

Comment:
i don't see a benefit of having distinction between return value 0 and positive. We managed to cancel alarm which was in the state that would result in expire callback or we canceled alarm which would not expire because it was not active (not started or already expired). We can check if we reach alarm by reading `k_cycle_get32()` and comparing it against `alarm->deadline`. Maybe we need `k_alarm_get_deadline()` API?

Reply:
>  i don't see a benefit of having distinction between return value 0 and positive.

And that's why counter cancellation doesn't provide this information and #12624/#12625 stay open.  Please understand that I do find knowing the state of the alarm to be useful, and it's not helpful to make me (as an application developer) compare values (one of which changes asynchronously) and try to reconstruct it when I could just have been given the information.

I've never needed `k_alarm_get_deadline()`.  Because this API doesn't use relative alarms it's always the value I last set it to, and I can read that from the `k_alarm` data structure.

Reply:
i agree that knowing the state of the alarm may be important but it's just that this information is not reliable because by the time i get the information deadline may already be met (but was not when alarm was canceled). maybe `k_alarm_cancel` should return:
- `-EINVAL` if alarm is not `SCHEDULED` or `ACTIVE`
- number of ticks before deadline when alarm was canceled

That would be more or less the same but positive value would have additional information.
Note that returning `0` may only happen in higher priority interrupt. Number of ticks left would allow conditions like `if ((u32_t)ret < 2) alarm_reached();`

Reply:
The alarm is also observably `READY` when multiple alarms reach their deadline in the same timer ISR, and one callback tries to cancel another alarm while it's still on the queue.

I think this can be resolved by clarifying that "reached its deadline" does not just mean that the RTC counter has advanced to the deadline (which we can't easily observe), but that the alarm and timer infrastructure has recognized that event has happened (which we can observe, because it's reflected in the alarm state being `READY`).

I will reword the documentation to make that clear.

This relates to your comment about the return value from `k_alarm_cancel()`.  As a user what I want to know is what the *state* of the alarm was (distinguish `SCHEDULED` from `READY`), not a difference in ticks that was captured at some arbitrary point in the past.  Any such difference is irrelevant since we don't know how long ago it was calculated.

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/16398#discussion_r288074327.

Can you provide a use case where having the time remaining at the point of cancellation provides more value than simply knowing the state of the alarm when it was cancelled, given that you don't know when the cancellation occurred?

Reply:
I will ask the similar question 😄 What is the use case for knowing that alarm infrastructure marked alarm as ready. If cancellation returns positive value, how can you be sure that it did not expire by the time you process that information (you may be interrupted for multiple ticks). If it is vital to do something special when cancelled alarm more or less reached deadline (but was cancelled before callback) you can lock, get `now` compare it with deadline and if it is close (user decides how close) you do something.